File: hdf-diff/src/fold-events.ts
import { formatTimestamp, parseTimestamp } from '@mitre/hdf-utilities';
import { computeSummary } from './summary.js';
import { computeEffectiveStatus } from './status.js';
import { computeFieldChanges, extractDrift, DEFAULT_TRACKED_FIELDS } from './diff.js';
import { dedupEvents, groupEventChains, verifyEventChain } from './event-stream-common.js';
import type { ApplyWarning } from './apply-events.js';
import type { RequirementDiff, ChangeReason, RequirementState } from './types.js';

type Doc = Record<string, unknown>;
type FieldChangesInput = Parameters<typeof computeFieldChanges>[0];

export interface FoldResult {
  comparison: Record<string, unknown>;
  warnings: ApplyWarning[];
}

function findSeedRequirement(doc: Doc, id: string): Doc | null {
  const baselines = (doc['baselines'] as Doc[] | undefined) ?? [];
  for (const baseline of baselines) {
    const requirements = (baseline['requirements'] as Doc[] | undefined) ?? [];
    const hit = requirements.find((r) => r['id'] === id);
    if (hit) return hit;
  }
  return null;
}

/**
 * Materialize a change-event batch into a full systemDrift hdf-comparison
 * against the seed document (ADR-0005 §5): each winning event lifts into the
 * Requirement_Diff shape the batch engine produces — before from the seed,
 * after from the event payload, field changes derived by the batch engine's
 * own computeFieldChanges — under the same fold contract as
 * applyChangeEvents (dedup, last-value-wins by sequence, tombstone-aware,
 * warnings never abort). Deterministic: the comparison timestamp is the
 * latest event occurrence, and seed-side override expiry anchors to the
 * seed's own timestamp (falling back to the event occurrence when the seed
 * has none) — never wall clock. Go parity: FoldChangeEventsIntoComparison.
 */
export async function foldChangeEventsIntoComparison(
  seedJson: string,
  events: Doc[],
): Promise<FoldResult> {
  const doc = JSON.parse(seedJson) as Doc;
  const docTimestamp = (doc['timestamp'] as string | undefined) ?? '';

  const { byKey, keys } = groupEventChains(dedupEvents(events));

  const warnings: ApplyWarning[] = [];
  const diffs: RequirementDiff[] = [];
  let maxOccurredMs = Number.NEGATIVE_INFINITY;
  let asOfFromEvents = '';
  let systemRef = '';

  for (const id of keys) {
    const chain = byKey.get(id) ?? [];
    const winner = chain[chain.length - 1];
    if (!winner) continue;
    if (systemRef === '') {
      systemRef = (winner['systemRef'] as string | undefined) ?? '';
    }

    let keyWarned = false;
    const keyWarn = (kind: string, message: string): void => {
      if (!keyWarned) {
        keyWarned = true;
        warnings.push({ requirementId: id, kind, message });
      }
    };

    const seedReq = findSeedRequirement(doc, id);
    const inSeed = seedReq !== null;
    await verifyEventChain(chain, seedReq, inSeed, keyWarn);

    // parseTimestamp + formatTimestamp keep the occurrence host-independent
    // and render timestamps as trimmed UTC, matching the Go peer's
    // winner.Timestamp.UTC().Format(RFC3339Nano).
    const occurred = parseTimestamp(winner['timestamp'] as string);
    if (occurred !== null && occurred.getTime() > maxOccurredMs) {
      maxOccurredMs = occurred.getTime();
      asOfFromEvents = formatTimestamp(occurred);
    }
    const newTs = occurred !== null ? formatTimestamp(occurred) : (winner['timestamp'] as string);
    // Seed-side override expiry needs a deterministic anchor: the seed's
    // own observation time, else the event occurrence — a timestamp-less
    // seed must never fall through to the wall clock.
    const seedRef = docTimestamp !== '' ? docTimestamp : newTs;

    if (winner['state'] === 'absent') {
      if (!seedReq) {
        keyWarn('absentUnknown', 'absent event for a key not present in the seed');
        continue;
      }
      diffs.push({
        id,
        title: seedReq['title'] as string | undefined,
        state: 'absent',
        oldEffectiveStatus: computeEffectiveStatus(seedReq, seedRef),
        changeReasons: [],
        oldImpact: seedReq['impact'] as number | undefined,
        fieldChanges: [],
        before: seedReq,
        after: null,
      });
      continue;
    }

    const after = winner['after'] as Doc | null;
    if (!after) {
      keyWarn('chainGap', 'winning event carries no applicable after payload');
      continue;
    }
    const newStatus = computeEffectiveStatus(after, newTs);

    // Content-bearing chain for a key the seed does not carry: the
    // comparison vocabulary allows a null before only on new, so the entry
    // is coerced to the batch engine's new shape; the warning records the
    // anomaly.
    if (!seedReq) {
      if (winner['state'] !== 'new') {
        keyWarn('chainGap', 'non-new chain for a key the seed does not carry; lifted as new');
      }
      diffs.push({
        id,
        title: after['title'] as string | undefined,
        state: 'new',
        newEffectiveStatus: newStatus,
        changeReasons: [],
        newImpact: after['impact'] as number | undefined,
        fieldChanges: [],
        before: null,
        after,
      });
      continue;
    }

    diffs.push({
      id,
      title: (after['title'] as string | undefined) ?? (seedReq['title'] as string | undefined),
      state: winner['state'] as RequirementState,
      oldEffectiveStatus: computeEffectiveStatus(seedReq, seedRef),
      newEffectiveStatus: newStatus,
      changeReasons: ((winner['changeReasons'] as string[] | undefined) ?? []) as ChangeReason[],
      oldImpact: seedReq['impact'] as number | undefined,
      newImpact: after['impact'] as number | undefined,
      fieldChanges: computeFieldChanges(
        seedReq as FieldChangesInput,
        after as FieldChangesInput,
        DEFAULT_TRACKED_FIELDS,
      ),
      before: seedReq,
      after,
    });
  }

  diffs.sort((a, b) => a.id.localeCompare(b.id));

  const asOf = asOfFromEvents !== '' ? asOfFromEvents : docTimestamp;
  if (asOf === '') {
    throw new Error(
      'cannot derive a comparison timestamp: no events supplied and the seed document has no timestamp',
    );
  }

  return {
    comparison: {
      formatVersion: '1.0.0',
      comparisonMode: 'systemDrift',
      timestamp: asOf,
      ...(systemRef !== '' ? { systemRef } : {}),
      sources: [
        { role: 'old', label: 'Old evaluation' },
        { role: 'new', label: 'New evaluation' },
      ],
      matching: { primaryStrategy: 'exactId' },
      summary: computeSummary(diffs),
      baselineDiffs: [],
      requirementDiffs: diffs,
      drift: extractDrift(diffs),
    },
    warnings,
  };
}

File: hdf-diff/go/fold_events.go
package diff

import (
	"encoding/json"
	"fmt"
	"sort"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
)

// FoldResult is the materialization output: a systemDrift comparison built
// from a change-event batch, plus any per-key anomalies detected while
// folding (same warning shape as ApplyChangeEvents).
type FoldResult struct {
	Comparison HdfComparison
	Warnings   []ApplyWarning
}

// eventStateToDiffState maps the event vocabulary onto the comparison
// vocabulary (values are identical; the types differ).
func eventStateToDiffState(s hdf.EventRequirementState) RequirementState {
	return RequirementState(string(s))
}

// eventReasonsToDiffReasons casts the event's producer-subset reasons onto
// the comparison vocabulary (the subset is value-identical by construction,
// test-enforced in hdf-schema).
func eventReasonsToDiffReasons(reasons []hdf.EventChangeReason) []ChangeReason {
	out := make([]ChangeReason, 0, len(reasons))
	for _, r := range reasons {
		out = append(out, ChangeReason(string(r)))
	}
	return out
}

// typedRequirement converts a map-decoded or event-carried requirement to
// the generated type; ok=false when untypeable.
func typedRequirement(v interface{}) (hdf.EvaluatedRequirement, bool) {
	var typed hdf.EvaluatedRequirement
	raw, err := json.Marshal(v)
	if err != nil {
		return typed, false
	}
	if err := json.Unmarshal(raw, &typed); err != nil {
		return typed, false
	}
	return typed, true
}

// nonNilFieldChanges guards the schema's required-array contract (nil
// marshals as null; Requirement_Diff.fieldChanges must be an array).
func nonNilFieldChanges(fc []FieldChange) []FieldChange {
	if fc == nil {
		return []FieldChange{}
	}
	return fc
}

// nonNilDiffs guards the schema's array contract for diff slices.
func nonNilDiffs(d []RequirementDiff) []RequirementDiff {
	if d == nil {
		return []RequirementDiff{}
	}
	return d
}

// FoldChangeEventsIntoComparison materializes a change-event batch into a
// full systemDrift hdf-comparison against the seed document (ADR-0005 §5):
// each winning event lifts into the Requirement_Diff shape the batch engine
// produces — before from the seed, after from the event payload, field
// changes derived by the batch engine's own computeFieldChanges — under the
// same fold contract as ApplyChangeEvents (dedup, last-value-wins by
// sequence, tombstone-aware, warnings never abort). Output is deterministic:
// the comparison timestamp is the latest event occurrence, and seed-side
// override expiry anchors to the seed's own timestamp (falling back to the
// event occurrence when the seed has none) — never the wall clock.
func FoldChangeEventsIntoComparison(seed []byte, events []*hdf.HDFRequirementChangeEvent) (*FoldResult, error) {
	var doc map[string]interface{}
	if err := json.Unmarshal(seed, &doc); err != nil {
		return nil, fmt.Errorf("failed to parse seed results: %w", err)
	}
	docTimestamp, _ := doc["timestamp"].(string)

	byKey, keys := groupEventChains(dedupEvents(events))

	result := &FoldResult{}
	keyWarnFactory := func(id string) func(kind, message string) {
		warned := false
		return func(kind, message string) {
			if !warned {
				warned = true
				result.Warnings = append(result.Warnings, ApplyWarning{RequirementID: id, Kind: kind, Message: message})
			}
		}
	}

	// Baseline names for the Baseline field: seed requirement's own baseline,
	// or the first baseline for keys the seed does not carry.
	firstBaselineName := ""
	baselineByReq := map[string]string{}
	if baselines, ok := doc["baselines"].([]interface{}); ok {
		for i, bRaw := range baselines {
			b, ok := bRaw.(map[string]interface{})
			if !ok {
				continue
			}
			name, _ := b["name"].(string)
			if i == 0 {
				firstBaselineName = name
			}
			reqs, _ := b["requirements"].([]interface{})
			for _, rRaw := range reqs {
				if r, ok := rRaw.(map[string]interface{}); ok {
					if id, _ := r["id"].(string); id != "" {
						baselineByReq[id] = name
					}
				}
			}
		}
	}

	var diffs []RequirementDiff
	var maxOccurred time.Time
	systemRef := ""

	for _, id := range keys {
		chain := byKey[id]
		winner := chain[len(chain)-1]
		if systemRef == "" {
			systemRef = winner.SystemRef
		}
		keyWarn := keyWarnFactory(id)

		var seedTyped *hdf.EvaluatedRequirement
		inSeed := false
		if _, reqs, idx, ok := findRequirement(doc, id); ok {
			if typed, tok := typedRequirement(reqs[idx]); tok {
				seedTyped = &typed
				inSeed = true
			}
		}

		verifyEventChain(chain, seedTyped, inSeed, keyWarn)

		if winner.Timestamp.After(maxOccurred) {
			maxOccurred = winner.Timestamp
		}
		newTs := winner.Timestamp.UTC().Format(time.RFC3339Nano)
		// Seed-side override expiry needs a deterministic anchor: the seed's
		// own observation time, else the event occurrence — a timestamp-less
		// seed must never fall through to the wall clock.
		seedRef := docTimestamp
		if seedRef == "" {
			seedRef = newTs
		}

		if winner.State == hdf.EventRequirementStateAbsent {
			if !inSeed {
				keyWarn("absentUnknown", "absent event for a key not present in the seed")
				continue
			}
			oldStatus := ComputeEffectiveStatus(*seedTyped, seedRef)
			oldImpact := seedTyped.Impact
			diffs = append(diffs, RequirementDiff{
				ID:                 id,
				Title:              resolveTitle(seedTyped.Title, nil),
				Baseline:           baselineByReq[id],
				State:              StateAbsent,
				OldEffectiveStatus: oldStatus,
				ChangeReasons:      []ChangeReason{},
				OldImpact:          &oldImpact,
				FieldChanges:       []FieldChange{},
				Before:             seedTyped,
				After:              nil,
			})
			continue
		}

		afterTyped, tok := typedRequirement(winner.After)
		if !tok {
			keyWarn("chainGap", "winning event carries no typeable after payload")
			continue
		}
		newStatus := ComputeEffectiveStatus(afterTyped, newTs)
		newImpact := afterTyped.Impact

		// Content-bearing chain for a key the seed does not carry: the
		// comparison vocabulary allows a null before only on new, so the
		// entry is coerced to the batch engine's new shape; the chain
		// warning (already emitted by verifyEventChain for anchored chains,
		// or here for unanchored ones) records the anomaly.
		if !inSeed {
			if winner.State != hdf.EventRequirementStateNew {
				keyWarn("chainGap", "non-new chain for a key the seed does not carry; lifted as new")
			}
			diffs = append(diffs, RequirementDiff{
				ID:                 id,
				Title:              resolveTitle(nil, afterTyped.Title),
				Baseline:           firstBaselineName,
				State:              StateNew,
				NewEffectiveStatus: newStatus,
				ChangeReasons:      []ChangeReason{},
				NewImpact:          &newImpact,
				FieldChanges:       []FieldChange{},
				Before:             nil,
				After:              &afterTyped,
			})
			continue
		}

		oldStatus := ComputeEffectiveStatus(*seedTyped, seedRef)
		oldImpact := seedTyped.Impact
		diffs = append(diffs, RequirementDiff{
			ID:                 id,
			Title:              resolveTitle(seedTyped.Title, afterTyped.Title),
			Baseline:           baselineByReq[id],
			State:              eventStateToDiffState(winner.State),
			OldEffectiveStatus: oldStatus,
			NewEffectiveStatus: newStatus,
			ChangeReasons:      eventReasonsToDiffReasons(winner.ChangeReasons),
			OldImpact:          &oldImpact,
			NewImpact:          &newImpact,
			FieldChanges:       nonNilFieldChanges(computeFieldChanges(*seedTyped, afterTyped, defaultTrackedFields)),
			Before:             seedTyped,
			After:              &afterTyped,
		})
	}

	sort.Slice(diffs, func(i, j int) bool { return diffs[i].ID < diffs[j].ID })
	if diffs == nil {
		diffs = []RequirementDiff{}
	}

	asOf := docTimestamp
	if !maxOccurred.IsZero() {
		asOf = maxOccurred.UTC().Format(time.RFC3339Nano)
	}
	if asOf == "" {
		return nil, fmt.Errorf("cannot derive a comparison timestamp: no events supplied and the seed document has no timestamp")
	}

	result.Comparison = HdfComparison{
		FormatVersion:    "1.0.0",
		ComparisonMode:   ModeSystemDrift,
		Timestamp:        asOf,
		SystemRef:        systemRef,
		Sources:          buildSources(ModeSystemDrift),
		Matching:         &MatchingConfig{PrimaryStrategy: "exactId"},
		Summary:          ComputeSummary(diffs),
		BaselineDiffs:    []BaselineDiff{},
		RequirementDiffs: diffs,
		Drift:            nonNilDiffs(extractDrift(diffs)),
	}
	return result, nil
}
